hydromt/components/mesh.py

```diff
@@ -12,33 +12,376 @@
+        res: float
+            Resolution used to generate 2D mesh [unit of the CRS], required if region
+            is not based on 'mesh'.
+        crs : EPSG code, int, optional
+            Optional EPSG code of the model or "utm" to let hydromt find the closest projected CRS.
+            If None using the one from region, and else 4326.
+        grid_name : str, optional
+            Name of the 2D grid in mesh, by default "mesh2d".
+
+        Returns
+        -------
+        mesh2d : xu.UgridDataset
+            Generated mesh2d.
+
+        """  # noqa: E501
+        self._logger.info("Preparing 2D mesh.")
+
+        # Create mesh2d
+        mesh2d = create_mesh2d(
+            region=region,
+            res=res,
+            crs=crs,
+            logger=self._logger,
+        )
+        # Add mesh2d to self.mesh
+        self.set(mesh2d, grid_name=grid_name)
+
+        # This setup method returns mesh2d so that it can be wrapped for models
+        # which require more information
+        return mesh2d
+
+    @property
+    def data(self) -> Union[xu.UgridDataArray, xu.UgridDataset]:
+        """
+        Model static mesh data. It returns a xugrid.UgridDataset.
+
+        Mesh can contain several grids (1D, 2D, 3D) defined according
+        to UGRID conventions. To extract a specific grid, use get_mesh
+        method.
+        """
+        # XU grid data type Xarray dataset with xu sampling.
+        if self._data is None and self._root.is_reading_mode():
+            self.read()
+        return self._data
+
+    @property
+    def crs(self) -> Optional[CRS]:
+        """Returns model mesh crs."""
+        if self.data is not None:
+            grid_crs = self.data.ugrid.crs
+            # Check if all the same
+            crs = None
+            for _k, v in grid_crs.items():
+                if crs is None:
+                    crs = v
+                if v == crs:
+                    continue
+                else:
+                    raise ValueError(
+                        f"Mesh crs is not uniform, please check {grid_crs}"
+                    )
+            return crs
+        else:
+            return None
+
+    @property
+    def bounds(self) -> Optional[Dict]:
+        """Returns model mesh bounds."""
+        if self.data is not None:
+            return self.data.ugrid.bounds
+        return None
+    @property 
+    def mesh_region(self) -> Optional[gpd.GeoDataFrame]:
+        """Return mesh total_bounds as a geodataframe."""
+        if self.data is not None:
+            region = gpd.GeoDataFrame(
+                geometry=[box(*self.data.ugrid.total_bounds)], crs=self.crs
+            )
+            return region
+        return None
+    @property
+    def mesh_names(self) -> List[str]:
+        """List of grid names in mesh."""
+        if self.data is not None:
+            return [grid.name for grid in self.data.ugrid.grids]
+        else:
+            return []
```

Comment:
Would be nice if your editor did autoformatting. We have enabled that in vs code settings. Perhaps you have some overwritten settings in your vs code personal settings. I expect some new lines in between

Reply:
Weird that it did not autoformat this time